[Origin API] Add more test coverage for opaque origin comparison

shannonbooth opened the pull request:
cc @mikewest @annevk 

Relating to https://github.com/whatwg/url/pull/892 and https://github.com/whatwg/html/pull/12058

One case which I didn't figure out a nice way of covering is opaque origin Worker which is _not_ a data scheme URL

shannonbooth committed:
[Origin API] Add more test coverage for opaque origin comparison

shannonbooth commented on html/browsers/origin/api/origin-from-worker.window.js:
```diff
@@ -0,0 +1,31 @@
+  const worker = new Worker(dataURL);
+  worker.onmessage = e => {
+    assert_true(e.data.isOpaque, "Origin created from data URL Worker should be an opaque origin.");
+    assert_false(e.data.sameOrigin, "Two data URL opaque origins should not be same-origin with one another.");
```
My original interpretation of https://html.spec.whatwg.org/multipage/workers.html#set-up-a-worker-environment-settings-object of returning a new opaque origin every time doesn't seem right, based on test results maybe I should make a MR to the HTML spec to change it to be more like the [Worklet](https://html.spec.whatwg.org/multipage/worklets.html#set-up-a-worklet-environment-settings-object) case which stores origin as a variable instead of creating a new one each time.

shannonbooth commented:
I pushed another commit to remove the file scheme URL test, as that has implementation defined behavior (discovered playing around making it not a tuple origin). Though actually, as I write this, maybe a test should instead be in a different file checking that:

```js
const fileSchemeIsOpaque = new URL('file:///something/').origin === 'null' ;
assert_eq(fileSchemeIsOpaque, Origin.from('file:///something/').opaque);
```

annevk reviewed (commented):
These are great @shannonbooth, thanks for writing them! I'll see if I need to change WebKit in a minute.

For `file:` URLs it seems reasonable to add `.tentative` test expecting an opaque origin.

I was hoping at some point we could at least agree that `file:` URL origins would always serialize as "`null`", but that didn't work out, but it might be time to try again.

shannonbooth commented:
> For `file:` URLs it seems reasonable to add `.tentative` test expecting an opaque origin.
> 
> I was hoping at some point we could at least agree that `file:` URL origins would always serialize as "`null`", but that didn't work out, but it might be time to try again.

Sure, added a tentative test to check that a file URL has an opaque origin. I'm not certain if the same origin behavior would tentatively be the same as all other opaque origin schemes, but I also added tests for that.

FWIW I am trying to make progress on moving towards opaque file origins in ladybird (https://github.com/LadybirdBrowser/ladybird/pull/7309 makes file origins opaque by default) to try and move closer to match other browsers behavior / security etc. The main point of investigation I have for this at the moment is on how to handle file scheme URLs in fetch, so hopefully I might be able to help making progress on https://github.com/whatwg/fetch/issues/1195 depending on how far I manage to progress 😄

annevk reviewed (approved):
Looks good modulo nit. Thanks again!

mikewest commented on html/browsers/origin/api/file-scheme-origin.tentative.any.js:
```diff
@@ -0,0 +1,14 @@
+
+test(t => {
+  const opaqueA = Origin.from("file:///path/to/a/file.txt");
+  const opaqueB = Origin.from("file:///path/to/a/file.txt");
```
Nit: I'd suggest extracting this string to a constant to make it clear that you're using the same URL in both origin constructions. I missed that on first skim.

```suggestion
  const fileURL = "file:///path/to/a/file.txt";
  const opaqueA = Origin.from(fileURL);
  const opaqueB = Origin.from(fileURL);
```

mikewest commented on html/browsers/origin/api/origin-comparison.any.js:
```diff
@@ -48,3 +48,15 @@ test(t => {
+
+test(t => {
+  const opaqueA = Origin.from("data:text/plain,opaque");
+  const opaqueB = Origin.from("data:text/plain,opaque");
```
Ditto here.

mikewest reviewed (approved):
LGTM too. I have small nits to improve clarity, but neither is important.
